altmetric/explorer/api/client.py

```diff
@@ -0,0 +1,80 @@
+import base64
+import hashlib
+import hmac
+
+import requests
+
+from .query import Query
+from .response import Response
+
+
+def digest(secret, message):
+    '''Calculates a cryptographic digest based on the user's API secret key and
+    the values of some of the parameters as described here:
+
+    https://www.altmetric.com/explorer/documentation/api#authentication
+    '''
+    if message is None:
+        message = ''
+    hmac_sha1 = hmac.new(secret.encode('utf-8'),
+                         message.encode('utf-8'), hashlib.sha1)
+    signature = base64.b64encode(hmac_sha1.digest()).decode('utf-8')
+    signature = hmac_sha1.hexdigest()
+    return signature
+
+
+class Client:
+    """Top level abstraction over the Altmetric Explorer API.
+    """
+
+    def __init__(self, api_endpoint, api_key, api_secret):
+        """Initialises a new Client object
+
+        Args:
+            api_endpoint (string): the url of the explorer api (usually https://www.altmetric.com/explorer/api)
+            api_key (string): your explorer api key
+            api_secret (string): your explorer api secret key
+
+        Raises:
+            ValueError: if the api key or the api secret is None
+
+        Notes:
+            You can find your api key and secret at https://www.altmetric.com/explorer/settings
+        """
+        if api_key == None or api_secret == None:
+            raise ValueError('api_key and api_secret cannot be None')
+
+        self.api_endpoint = api_endpoint
+        self.api_key = api_key
+        self.api_secret = api_secret
+
+    def get(self, path, **args):
+        """Generic get method that constructs a call to an API path and returns a Response. An authentication digest is calculated behind the scenes using the
+        api keys instance variables and the filters provided and added to the request automatically.
+
+        Args:
+            path (string): the path to query
+            args (keyword list, accepts the following):
+                page_size (int, optional): size of each page to be returned. Defaults to 100.
+                order (string, optional): the field on which the results should be sorted. Defaults to None.
+                limit (int, optional): maximum number of items to return.  Set to None to return everything. Defaults to None.
+
+                All other keyword arguments are treated as filters e.g. timeframe, mention_sources_countries
+
+        Returns:
+            response: A Response object for the API call.
+        """
+        query = Query(**args)
```

Comment:
I wonder if it would also be valuable to create some kind of API query string method to expose the actual URL with params that is being sent to Explorer API like:

```
api_client = Client('https://www.altmetric.com/explorer/api', API_KEY, API_SECRET)
api_client.get_mention_sources_request_url_query(timeframe='3d')

# "https://www.altmetric.com/explorer/api/research_outputs/mention_sources?digest=<API_SECRET>&key=<API_KEY>&timeframe='3d'"
```

A couple of benefits I can see:
1. Allows users and engineers to understand how the client is crafting the request and potentially debug when the client is not returning expected results by comparing it against the docs
2. Users can share their query string (sanitised of API keys/secrets of course) with other Explorer API users who do do not have access to/prefer not to use the API Client.

Reply:
Definitely valuable, but after a preliminary think of how best to implement it I think the cost might be too much. The best way to implement it would be to have something like below:

```py
client = Client('https://www.altmetric.com/explorer/api', API_KEY, API_SECRET)
response = client.get('your/url/of/choice', timeframe='3d', q='foo')
print(response.url)  # => "https://www.altmetric.com/explorer/api/research_outputs/mention_sources?digest=<API_SECRET>&key=<API_KEY>&timeframe='3d'"
```

But that would require a fairly large refactor on Response to take a URL string in its initializer instead of an HTTP response, and perform the HTTP request during initialization.

An interesting idea, but beyond the scope of this feature in my opinion!

Reply:
As mentioned above, I like this to be a leaky abstraction so we have included a way to get the response from the HTTP client and you can use it to get the url.

```python
response = client.get('your/url/of/choice', timeframe='3d', q='foo')
print(response.raw_response.url)
```

This is described in the [getting started notebook](https://github.com/altmetric/altmetric-explorer-api-client/pull/1/files#diff-09515ed158b8331c13c29aa5f5e03884d2cfdf62b1f5c7eefef1ddb4d26383ceR279-R295).

I don't see a massive problem in making a `url` method on `Response` to make this a bit cleaner but at the same time the big unknown here is how people will actually use the client in real life.  We do already have the official API docs that explain how URLs are built and the recommended way to use the API is to query in Explorer and then and then grab the API query from the `Open in API` menu item.

![image](https://github.com/altmetric/altmetric-explorer-api-client/assets/249602/9b005aa2-e1ff-4e32-8131-241a3858877d)